### DISCUSSION TITLE
Hole punching with Open Relay for file transfer

### DISCUSSION THREAD

**Discussion by LeCyberDucky** [libp2p/rust-libp2p#4266] [Q&A] [answered]
Hey there!

For some background, I am not new to Rust, but I am fairly new to networking. 
I want to create a project that will allow me to send a file from my computer to my friends computer. After some fiddling around with TCP, I have come to the conclusion that establishing a direct connection between two computers sitting behind each their router is not that straight forward. 

I believe that I will need to perform hole punching to set up this direct connection. I have read these articles: 
https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/
https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html

The tutorial assumes that I have access to some public relay server that I can run my binary on. This leads me to my first question: 
Can I use the [Open Relay Project](https://www.metered.ca/tools/openrelay/) for my relay server? I assume that the idea with these servers is that they already provide the wanted functionality, and I can't just upload my own binaries to these servers. But can I configure my libp2p program to work with these servers?  

If this is possible, how would you suggest me to get started? I would like to create a minimal example of just sending a string from one computer to the other. Is one of the [examples](https://github.com/libp2p/rust-libp2p/tree/master/examples) a particularly good starting point for this? I have seen the [ping tutorial](https://docs.rs/libp2p/latest/libp2p/tutorials/ping/index.html), but will that help me for communicating across networks?

**Comment by thomaseizinger** [comment, marked as answer]
Openrelay uses WebRTC TURN whereas libp2p has its own relay protocol so the two are not compatible.

Running a relay can be expensive depending on how much bandwidth you are offering to your users.

It is best if you run your own relay server. You can then use the dcutr protocol to attempt a hole punch between two peers. Checkout our examples on how to do that :)

**Reply by LeCyberDucky** [reply]
Hmm, I see. Running my own server doesn't suit my use case very well, but thanks for the quick response!
